File: 14-Stack-and-Queue-Part-II/cpp/03-lfu-cache.cpp
class LFUCache {
    struct Node {
        int key, val, freq;
        Node *prev, *next;
        Node(int k, int v) : key(k), val(v), freq(1), prev(nullptr), next(nullptr) {}
    };

    struct DoublyLinkedList {
        Node *left, *right;
        int size;

        DoublyLinkedList() {
            left = new Node(0, 0);
            right = new Node(0, 0);
            left->next = right;
            right->prev = left;
            size = 0;
        }

        void pushRight(Node* node) {
            Node* p = right->prev;
            p->next = node;
            node->prev = p;
            node->next = right;
            right->prev = node;
            size++;
        }

        void pop(Node* node) {
            Node* p = node->prev;
            Node* n = node->next;
            p->next = n;
            n->prev = p;
            size--;
        }

        Node* popLeft() {
            Node* node = left->next;
            pop(node);
            return node;
        }
    };

    int capacity;
    int lfuCount;
    unordered_map<int, Node*> nodeMap;
    unordered_map<int, DoublyLinkedList*> listMap;

public:
    LFUCache(int capacity) : capacity(capacity), lfuCount(0) {}

    int get(int key) {
        if (nodeMap.find(key) == nodeMap.end()) return -1;
        Node* node = nodeMap[key];
        counter(node);
        return node->val;
    }

    void put(int key, int value) {
        if (capacity == 0) return;

        if (nodeMap.find(key) != nodeMap.end()) {
            Node* node = nodeMap[key];
            node->val = value;
            counter(node);
            return;
        }

        if ((int)nodeMap.size() == capacity) {
            Node* toRemove = listMap[lfuCount]->popLeft();
            nodeMap.erase(toRemove->key);
            delete toRemove;
        }

        Node* node = new Node(key, value);
        nodeMap[key] = node;
        if (listMap.find(1) == listMap.end()) {
            listMap[1] = new DoublyLinkedList();
        }
        listMap[1]->pushRight(node);
        lfuCount = 1;
    }

private:
    void counter(Node* node) {
        int count = node->freq;
        listMap[count]->pop(node);

        if (count == lfuCount && listMap[count]->size == 0) {
            lfuCount++;
        }

        node->freq++;
        if (listMap.find(node->freq) == listMap.end()) {
            listMap[node->freq] = new DoublyLinkedList();
        }
        listMap[node->freq]->pushRight(node);
    }
};

File: 14-Stack-and-Queue-Part-II/python/03_lfu_cache.py
from collections import defaultdict

class Node:
    def __init__(self, key: int = 0, val: int = 0):
        self.key = key
        self.val = val
        self.freq = 1
        self.prev = None
        self.next = None

class DoublyLinkedList:
    def __init__(self):
        self.left = Node()
        self.right = Node()
        self.left.next = self.right
        self.right.prev = self.left
        self.size = 0

    def push_right(self, node):
        prev = self.right.prev
        prev.next = node
        node.prev = prev
        node.next = self.right
        self.right.prev = node
        self.size += 1

    def pop(self, node):
        prev, nxt = node.prev, node.next
        prev.next = nxt
        nxt.prev = prev
        node.prev = None
        node.next = None
        self.size -= 1

    def pop_left(self):
        node = self.left.next
        self.pop(node)
        return node

class LFUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.lfu_count = 0
        self.node_map = {}
        self.list_map = defaultdict(DoublyLinkedList)

    def _counter(self, node):
        count = node.freq
        self.list_map[count].pop(node)

        if count == self.lfu_count and self.list_map[count].size == 0:
            self.lfu_count += 1

        node.freq += 1
        self.list_map[node.freq].push_right(node)

    def get(self, key: int) -> int:
        if key not in self.node_map:
            return -1
        node = self.node_map[key]
        self._counter(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        if self.capacity == 0:
            return

        if key in self.node_map:
            node = self.node_map[key]
            node.val = value
            self._counter(node)
            return

        if len(self.node_map) == self.capacity:
            to_remove = self.list_map[self.lfu_count].pop_left()
            del self.node_map[to_remove.key]

        node = Node(key, value)
        self.node_map[key] = node
        self.list_map[1].push_right(node)
        self.lfu_count = 1
